# Biodiversity in National Parks - Data Analysis
### Analysis Overview
In this analysis, we will be analyzing data from two datasets containing information on National Parks and the species who inhabit them.

The datasets are:

- ```"observations.csv"``` - includes data on the national park being observed, what species live their, and how many times each species has been recorded being spotted within the last 7 days.
- ```"species_info.csv"``` - includes data on a large number of different species.  It features the species' scientific name, common name, category of species, and conservation status.

---

## Section 1 - Understanding the Data
In order to work with any data, you first need to understand what you'll be working with.  

1. Import the Python libraries you'll need to conduct your analysis 
2. Load the two CSV datasets into two seperate Pandas DataFrames, ```parks``` and ```species```
3. View the first 10 rows of the ```parks``` dataframe
4. View the first 10 rows of the ```species``` dataframe

In [259]:
# 1

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

In [260]:
# 2

parks = pd.read_csv('observations.csv')
species = pd.read_csv('species_info.csv')

In [261]:
# 3 

parks.head(10)

scientific_name                            park_name  \
0                 Vicia benghalensis  Great Smoky Mountains National Park   
1                     Neovison vison  Great Smoky Mountains National Park   
2                  Prunus subcordata               Yosemite National Park   
3               Abutilon theophrasti                  Bryce National Park   
4           Githopsis specularioides  Great Smoky Mountains National Park   
5  Elymus virginicus var. virginicus               Yosemite National Park   
6                   Spizella pusilla            Yellowstone National Park   
7                  Elymus multisetus  Great Smoky Mountains National Park   
8             Lysimachia quadrifolia               Yosemite National Park   
9         Diphyscium cumberlandianum            Yellowstone National Park   

   observations  
0            68  
1            77  
2           138  
3            84  
4            85  
5           112  
6           228  
7            39  
8           168  
9           250

In [262]:
# 4

species.head(10)

category                scientific_name  \
0   Mammal  Clethrionomys gapperi gapperi   
1   Mammal                      Bos bison   
2   Mammal                     Bos taurus   
3   Mammal                     Ovis aries   
4   Mammal                 Cervus elaphus   
5   Mammal         Odocoileus virginianus   
6   Mammal                     Sus scrofa   
7   Mammal                  Canis latrans   
8   Mammal                    Canis lupus   
9   Mammal                    Canis rufus   

                                        common_names conservation_status  
0                           Gapper's Red-Backed Vole                 NaN  
1                              American Bison, Bison                 NaN  
2  Aurochs, Aurochs, Domestic Cattle (Feral), Dom...                 NaN  
3  Domestic Sheep, Mouflon, Red Sheep, Sheep (Feral)                 NaN  
4                                      Wapiti Or Elk                 NaN  
5                                  White-Tailed Deer                 NaN  
6                                Feral Hog, Wild Pig                 NaN  
7                                             Coyote  Species of Concern  
8                                          Gray Wolf          Endangered  
9                                           Red Wolf          Endangered

The first thing that stands out to me is in the ```species``` dataframe.  Without looking any deeper you can already see that it contains some missing values ('NaN').  In order to accurately work with this data, you'll need to clean up your dataframes.  Aside from that one column in the ```species``` dataframe, you aren't sure which columns may have missing data in either of the two dataframes, so you'll need to do some work to find out.

Let's focus on looking at your ```parks``` dataframe first!

5. Check to see if the ```parks``` dataframe contains any missing values.  Check column names and data types as well
6. Take a look at the summary statistics
7. find out how many species observations there are for each National Park 

In [263]:
# 5

parks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23296 entries, 0 to 23295
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   scientific_name  23296 non-null  object
 1   park_name        23296 non-null  object
 2   observations     23296 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 546.1+ KB


In [264]:
# 6

parks.describe(include='all')

scientific_name                            park_name  observations
count              23296                                23296  23296.000000
unique              5541                                    4           NaN
top     Myotis lucifugus  Great Smoky Mountains National Park           NaN
freq                  12                                 5824           NaN
mean                 NaN                                  NaN    142.287904
std                  NaN                                  NaN     69.890532
min                  NaN                                  NaN      9.000000
25%                  NaN                                  NaN     86.000000
50%                  NaN                                  NaN    124.000000
75%                  NaN                                  NaN    195.000000
max                  NaN                                  NaN    321.000000

In [265]:
# 7

parks.park_name.value_counts()

Great Smoky Mountains National Park    5824
Yosemite National Park                 5824
Bryce National Park                    5824
Yellowstone National Park              5824
Name: park_name, dtype: int64

You can see that the length of the dataframe is 23,296 rows long and that it has 3 columns, each with 0 missing values.  The column names are clean and easy to work with and you can also see the data types of each column make sense.

Looking at the output from code block #6, you can see that there are a lot of unique species within the dataframe (5541), but only 4 diferent National Parks.    

Since there isn't a lot of unique values, you can easily use ```.value_counts()``` on the dataframe to show that there is 5824 observations of species in each National Park.  

Though our data looks seemingly nice and tidy, we could further inspect the ```scientific_name``` column to see if there are any strange values that are hidden within the column that could effect the analysis.  An example of a strange value could be an empty string ("") or a string that says ```"NULL"``` or ```"NaN"```.  These would technically count as values but they should instead be classified as a missing value.

8. Pull some random samples of the ```scientific_name``` column, look over the values, and check to see if there are any entries that don't belong
  

In [266]:
# 8

for num in range(10):
    print(f'Sample {num + 1}')
    parks_sample = parks.scientific_name.sample(25)
    print(parks_sample)
    print()

Sample 1
21010                Sphyrapicus nuchalis
12083                 Castilleja cusickii
15122                     Smilax herbacea
11836               Populus grandidentata
13316           Forsstroemia trichomitria
2959                Sphenopholis obtusata
18299           Thryomanes bewickii altus
15976                  Ludwigia decurrens
5864                          Xyris torta
10511                  Smilax biltmoreana
11313             Ranunculus occidentalis
20650                     Trillium simile
13337                   Scleria oligantha
16677              Cephalozia lunulifolia
10046    Oxytropis deflexa var. foliolosa
1301                   Collomia tinctoria
14186                       Arabis lyrata
7985                      Camelina sativa
5371                       Carex capitata
21646             Allophyllum divaricatum
19330                         Canis lupus
3840                    Picoides villosus
15996               Platanus occidentalis
17128                Ranu

According to the samples, the data is good to go!

Time to check out the ```species``` dataframe!

9. Check to see if the ```species``` dataframe contains any missing values.  Check column names and data types as well
10. Take a look at the summary statistics
11. Find out how many species there are within each ```category```
12. Find out how many species there are within each ```conservation_status```
13. Find out the number of ```NaN``` values in the ```conservation_status``` column, and the percentage

In [267]:
# 9 

species.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5824 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5824 non-null   object
 1   scientific_name      5824 non-null   object
 2   common_names         5824 non-null   object
 3   conservation_status  191 non-null    object
dtypes: object(4)
memory usage: 182.1+ KB


In [268]:
# 10

species.describe(include='all')

category    scientific_name        common_names  \
count             5824               5824                5824   
unique               7               5541                5504   
top     Vascular Plant  Castor canadensis  Brachythecium Moss   
freq              4470                  3                   7   

       conservation_status  
count                  191  
unique                   4  
top     Species of Concern  
freq                   161

In [269]:
# 11

species.category.value_counts()

Vascular Plant       4470
Bird                  521
Nonvascular Plant     333
Mammal                214
Fish                  127
Amphibian              80
Reptile                79
Name: category, dtype: int64

In [270]:
# 12

species.conservation_status.value_counts()

Species of Concern    161
Endangered             16
Threatened             10
In Recovery             4
Name: conservation_status, dtype: int64

In [271]:
# 13

nan_count = len(species[species.conservation_status.isnull() == True])
nan_percent = (len(species[species.conservation_status.isnull() == True]) / len(species)) * 100

print(f'NaN values in \'conservation_status\' column: {nan_count} out of {len(species)}')
print(f'Percentage of NaN values in \'conservation_status\' column: {round(nan_percent, 2)}%')

NaN values in 'conservation_status' column: 5633 out of 5824
Percentage of NaN values in 'conservation_status' column: 96.72%


Going through the summary statistics gives us plenty of insight into the data.  The ```conservation_status``` column is clearly missing a massive chunk of data,
96.72% actually.  But why is this data missing?  If you take a look at the unique values within the column, you can see that there are only 4 values, and all of them are associated with endangerment to some degree (output of code block #12).  With this knowledge, we can assume that the ```NaN``` values just mean that the species is not at risk of endangerment.  The values seem to be missing for a reason.

With this assumption, you can go ahead and fill the missing values with a value that signifies "no risk".

14. Fill in the ```NaN``` values in the ```conservation_status``` column with the string value ```"Not at risk"```

In [272]:
# 14

species.conservation_status.fillna('Not at Risk', inplace=True)
species.conservation_status.value_counts()

Not at Risk           5633
Species of Concern     161
Endangered              16
Threatened              10
In Recovery              4
Name: conservation_status, dtype: int64

After doing this, it wouldn't hurt to check the scientific name column for any duplicate values.  If you refer back to code block #10, you can see that the column does have some repeating values.  This should be looked into.

15. Figure out how many values are duplicates within the ```scientific_name``` column
16. Find out how many duplicate rows exist within the ```species``` dataframe, if any

In [273]:
# 15

dup = species.scientific_name.duplicated().sum()
print(f'Duplicated values in \'scientific_name\' column: {dup}')

Duplicated values in 'scientific_name' column: 283


In [274]:
# 16 

dup = species.duplicated().sum()
print(f'Duplicated rows in \'species\' dataframe: {dup}')

Duplicated rows in 'species' dataframe: 0


So there aren't any duplicate rows within the ```species``` dataframe, but there are 283 duplicates within the ```scientific_name``` column of the ```species``` dataframe.  This may be due to the fact that more than one person recorded the same species and they recorded different numbers observed in the ```observations``` column.  Let's find out.

17. Experiment with the ```species``` dataframe and determine if these duplicate values should be dropped of not

In [275]:
# 17

species[species.scientific_name.duplicated()].head()

category         scientific_name                          common_names  \
3017   Mammal          Cervus elaphus                    Rocky Mountain Elk   
3019   Mammal  Odocoileus virginianus  White-Tailed Deer, White-Tailed Deer   
3020   Mammal             Canis lupus                       Gray Wolf, Wolf   
3022   Mammal           Puma concolor           Cougar, Mountain Lion, Puma   
3025   Mammal        Lutra canadensis                           River Otter   

     conservation_status  
3017         Not at Risk  
3019         Not at Risk  
3020         In Recovery  
3022         Not at Risk  
3025         Not at Risk

In [276]:
# 17 (Cont.)

species[species.scientific_name == 'Canis lupus']

category scientific_name     common_names conservation_status
8      Mammal     Canis lupus        Gray Wolf          Endangered
3020   Mammal     Canis lupus  Gray Wolf, Wolf         In Recovery
4448   Mammal     Canis lupus  Gray Wolf, Wolf          Endangered

This doesn't seem right.  This is the same mammal, recorded 3 times, and once with a different conservation status.  The only sensible explanation for the repeated entries could be due to a possible change in conservation status.  If that is the case, then we can rest knowing it was recorded for a reason. 

In [277]:
# 17 (Cont.)

species[species.scientific_name == 'Cervus elaphus']

category scientific_name        common_names conservation_status
4      Mammal  Cervus elaphus       Wapiti Or Elk         Not at Risk
3017   Mammal  Cervus elaphus  Rocky Mountain Elk         Not at Risk

In [278]:
# 17 (Cont.)

species[species.scientific_name == 'Puma concolor']

category scientific_name                 common_names conservation_status
16     Mammal   Puma concolor      Panther (Mountain Lion)         Not at Risk
3022   Mammal   Puma concolor  Cougar, Mountain Lion, Puma         Not at Risk
4451   Mammal   Puma concolor                Mountain Lion         Not at Risk

Ok, no rest for us, but you can conclude that a changing conservation status is NOT the reason for the duplicate values, and this is comforting, sort of.  It is safe to assume that the duplicates within the dataframe were recorded by different people who each had slightly different common names for the species.  To accurately plot any data, you'll need to drop the duplicates since they will do nothing for you but scew your results ever so slightly.

18. Drop duplicates in the ```scientific_name``` column of the ```species``` dataframe

In [287]:
species.drop_duplicates(subset=['scientific_name'], inplace=True)
print(species.scientific_name.duplicated().sum())
print(len(species))
print(species.scientific_name.nunique())

0
5541
5541


18. Find out how many duplicate rows exist within the ```parks``` dataframe, if any
19. Experiment with the ```parks``` dataframe and determine if these duplicate values should be dropped of not

In [280]:
# 18

parks.duplicated().sum()

15

In [281]:
# 19



# Appendix
### Section 1
1. Python library imports:
* Numpy - Used for fast calculations within pandas.  Can perform matrix calculations.
* Pandas - Used for loading in datasets into workable dataframes.  Crucial for Data Analysis with Python.
* Matplotlib - Used for creating plots and visualizations from pandas dataframes.
* Seaborn - Newer, more intuitive visualization library built to work with Matplotlib and pandas.
* Statsmodels - Used for creating predictive models.

2. Use the ```.read_csv()``` method along with ```pd``` (pandas) to import a csv file into a pandas dataframe.  The method takes the CSV filename surrounded by quotations as a parameter (Example: parks = pd.read_csv("observations.csv"))

3. Use the ```.head()``` method from the ```pandas``` library on the ```parks``` dataframe to show the first 10 rows of the dataframe

4. Use the ```.head()``` method from the ```pandas``` library on the ```species``` dataframe to show the first 10 rows of the dataframe

5. Use the ```.info()``` method from the ```pandas``` library on the ```parks``` dataframe to show column names, data types of the columns, length of the dataframe, and possible missing values within the dataframe.  This shows that the length of the dataframe is 23,296 rows long and that it has 3 columns, each with 0 missing values.  We can see that the column names are clean and easy to work with and we can also see the data types of each column, which are workable for now.

6. Use the ```.decribe()``` method from the ```pandas``` library on the ```parks``` dataframe to show some summary statistics about each column.  In order to see summary statistics for categorical variables, you need to use the ```include=``` parameter and set its value to ```'all'```.  This shows the amount of unique values for the categorical variables (```scientific_name``` and ```park_name```), and some useful statistics about the quantitative variable (```observations```).  The average number of times a single species was observed in 7 days was 142 times.  The minimum times observed was 9 times and the maximum was 321 times.  This is useful primarily because now you know that the ```observations``` column doesn't contain any zero values in it.  This means you can move on to exploring a different variable within the data.

7. Use the ```.value_counts()``` method from the ```pandas``` library on the ```parks.park_name``` dataframe column to show the frequency of the unique values within the column.  You can see that each of the 4 National Parks contain an equal amount of entries.  This information also shows that the 4 unique values within this column are in fact National Parks.

8. Create a ```for``` loop that loops 10 times.  With every loop print a string of the sample number.  With every loop you'll pull a random sample of the ```parks.scientific_name``` column using the pandas ```.sample()``` method.  The method will take an integer as a parameter which specifies the size of the random sample.  The larger the sample, the better.  After the samples are pulled, the samples should be examined for any values that are not scientific names.  If you pull enough samples of a decent size, then this can accurately portray the dataset as a whole.  This is much more efficient than reading through all 5541 unique scientific names.
